src/java/org/apache/cassandra/index/sai/disk/vector/OnDiskOrdinalsMap.java

```diff
@@ -241,21 +243,56 @@ public void close()
         }
     }
 
+    /**
+     * not thread safe
+     */
     private class FileReadingOrdinalsView implements OrdinalsView
     {
         RandomAccessReader reader = fh.createReader();
         private final long high = (segmentEnd - 8 - rowOrdinalOffset) / 8;
+        private int lastFoundRowId = -1;
+        private long lastFoundRowIdIndex = -1;
 
         /**
          * @return order if given row id is found; otherwise return -1
          */
         @Override
         public int getOrdinalForRowId(int rowId) throws IOException
         {
+            assert rowId >= lastFoundRowId : "rowId < lastFoundRowId, rowId " + rowId + " lastFoundRowId " + lastFoundRowId + " lastFoundRowIdIndex " + lastFoundRowIdIndex + " high " + high;
```

Comment:
Sorry, I think my terms might not have been clear earlier. I think we need to keep track of the last `rowId` passed to this method, and we should assert that the next `rowId` is greater. We use this pattern here:

https://github.com/datastax/cassandra/blob/c70301478f8aa03ba4be0369f0c71ca8aaede0fe/src/java/org/apache/cassandra/index/sai/disk/v1/bitpack/AbstractBlockPackedReader.java#L91-L97